### ISSUE TITLE
HttpClient does not work with Request.EnableRewind

### ISSUE THREAD

**Issue by Tratcher** [aspnet/AspNetCore#3476]
_From @cnblogs-dudu on August 21, 2018 9:39_

Perhaps it is `HttpConnection.WriteAsync` does not work with `FileBufferingReadStream`.

Website A with the following code is listening 5000.
```cs
public class HomeController : Controller
{
    public async Task<IActionResult> Index()
    {
        Request.EnableRewind();
        var streamContent = new StreamContent(Request.Body);
        var httpClient = new HttpClient();
        var response = await httpClient.PostAsync("http://localhost:5002", streamContent);
        return Ok(await response.Content.ReadAsStringAsync());
    }
}
```
Website B with the following code is listening 5002.
```cs
public class HomeController : Controller
{
    public async Task<IActionResult> Index()
    {
        using (var ms = new MemoryStream())
        {
            await Request.Body.CopyToAsync(ms);
            return Ok(ms.Length);
        }
    }
}
```
Use the following curl command to send a request.
```shell
curl -s -d 'test' localhost:5000/
```
The actual output is 0. The expected output is 4.

If remove `Request.EnableRewind()`, it works.
If put the following code between `Request.EnableRewind()` and `HttpClient` , it works.
```cs
using (var ms = new MemoryStream())
{
    await Request.Body.CopyToAsync(ms);
}
Request.Body.Seek(0, 0);
```

_Copied from original issue: dotnet/corefx#31866_

**Comment by Tratcher** [comment]
_From @davidsh on August 21, 2018 15:51_

cc: @Tratcher

**Comment by Tratcher** [comment]
_From @cnblogs-dudu on August 23, 2018 9:48_

I figured it out.  `Content-Length: 0` resulted in this issue. I will try to create a pull request.
```
POST / HTTP/1.1
Content-Length: 0
Host: localhost:5002


test
```

**Comment by Tratcher** [comment]
_From @cnblogs-dudu on August 23, 2018 11:32_

If change `return true` to `return length > 0` in `TryComputeLength` of `StreamContent`, 
```cs
protected internal override bool TryComputeLength(out long length)
{
    if (_content.CanSeek)
    {
        length = _content.Length - _start;
        return length > 0;
    }
    else
    {
        length = 0;
        return false;
    }
}
```
It works and gets the following http request.
```text
POST / HTTP/1.1
Transfer-Encoding: chunked
Host: localhost:5002

4
test
0



```

**Comment by Tratcher** [comment]
_From @davidsh on August 23, 2018 14:12_

>If change return true to return length > 0 in TryComputeLength of StreamContent,

This might fix your scenario. But this changes the original design of 'TryComputeLength'.  It was supposed to return true if it can compute a length regardless of what that value is (zero or greater than zero).

We'll need to study this more to see if this adjustment will be compatible.  The workaround, though, is for a developer to create a subclass of StreamContent and override the 'TryComputeLength' method as you have shown.

cc: @geoffkizer @karelz @stephentoub

**Comment by Tratcher** [comment]
_From @cnblogs-dudu on August 25, 2018 14:32_

`return length > 0` only works in the sample app. It does not works in our production scenario that putting file to cloud storage.

I fixed it in HttpAbstractions repo. I will submit a PR to HttpAbstractions.

**Comment by Tratcher** [comment]
What this comes down to is that Request.Body and StreamContent make fundamentally different assumptions and I don't think it makes sense to change either to accommodate the other.

Also, forwarding requests is a lot more involved than you imply above. See the prototype over at https://github.com/aspnet/proxy for some examples.

**Comment by Tratcher** [comment]
_From @davidfowl on August 29, 2018 18:26_

This scenario should work. Lets keep an issue open on the ASP.NET Core side of things.

**Comment by Eilon** [comment]
@Tratcher any idea what 'repo' this is part of?

**Comment by Tratcher** [comment]
HttpAbstractions. Marked.

**Comment by cnblogs-dudu** [comment]
The scenario is that we have a file uploading website that it receives uploading files and transfers them to Alibaba Cloud Object Storage Service and Tencent Cloud Object Storage.

**Comment by cnblogs-dudu** [comment]
As a Stream, FileBufferingReadStream  has two unexpected behaviours.
* First time reading the whole stream behaves differently from later reading.
* In first time reading, it's CanSeek is true and it's bytes can be read, but it's length is 0.

StreamContent treats FileBufferingReadStream  as Stream right. But FileBufferingReadStream is not a right Stream.

**Comment by Tratcher** [comment]
There are constraints in the Stream contract itself that contribute to this problem. The buffer stream is seekable forwards and backwards for it's whole lifetime. Stream says that if you implement CanSeek you must also implement Length. Unfortunately it doesn't know the real length, only the currently buffered length (Content-Length would provide a hint, but Chunked wouldn't). In this it's actually similar to MemoryStream and FileStream where the length can change during the lifetime of the stream. StreamContent is oversimplifying it's interpretation of Length but it doesn't have a lot of other choices.

In short, StreamContent and FileBufferingReadStream are both wrong, but are both as correct as possible with the data available.

**Comment by anurse** [comment]
We don't support flowing the Request Body through as a stream to HttpClient.
